From b0f12b0b05a7b32ee955f414b785c9169de9166a Mon Sep 17 00:00:00 2001
From: Harshesh Valera <harshesh.v@variscite.com>
Date: Fri, 18 Jan 2019 10:26:28 -0800
Subject: [PATCH] imx-test: mxc_v4l2_overlay: add autofocus support

Add -af switch to control autofocus support
-af <autofocus enable>
0 - Autofocus Disabled
1 - Autofocus Enabled (Default)
Signed-off-by: Harshesh Valera <harshesh.v@variscite.com>
---
 test/mxc_v4l2_test/mxc_v4l2_overlay.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/test/mxc_v4l2_test/mxc_v4l2_overlay.c b/test/mxc_v4l2_test/mxc_v4l2_overlay.c
index a1adbe6..f89824c 100644
--- a/test/mxc_v4l2_test/mxc_v4l2_overlay.c
+++ b/test/mxc_v4l2_test/mxc_v4l2_overlay.c
@@ -94,6 +94,7 @@ unsigned long loc_alpha_phy_addr1;
 int alpha_buf_size = 0;
 int ctrl_c_rev = 0;
 int g_fd_fb_fg = 0;
+int af_enabled = 1; /* Enable Auto Focus by default */
 
 static void print_pixelformat(char *prefix, int val)
 {
@@ -551,6 +552,9 @@ int process_cmdline(int argc, char **argv)
 		else if (strcmp(argv[i], "-a") == 0) {
 			g_alpha_mode = atoi(argv[++i]);
 		}
+		else if (strcmp(argv[i], "-af") == 0) {
+			af_enabled = atoi(argv[++i]);
+		}
                 else if (strcmp(argv[i], "-help") == 0) {
                         printf("MXC Video4Linux overlay Device Test\n\n" \
                                " -iw <input width>\n -ih <input height>\n" \
@@ -566,7 +570,8 @@ int process_cmdline(int argc, char **argv)
 			       " -a <alpha mode> 0-global alpha blending 1-local alpha blending\n" \
 			       " -fr <frame rate> 30fps by default\n" \
                                " -fg foreground mode when -fg specified,"
-                               " otherwise go to frame buffer\n");
+                               " otherwise go to frame buffer\n" \
+                               " -af <autofocus enable> 0-disabled 1-enabled, enabled by default\n");
                         return -1;
                 }
         }
@@ -600,6 +605,9 @@ main(int argc, char **argv)
 	int ret = 0;
 	struct sigaction act;
 	struct v4l2_dbg_chip_ident chip;
+	struct v4l2_control c;
+		c.id = V4L2_CID_FOCUS_AUTO;
+		c.value = 1;
 
 	print_name(argv);
 
@@ -617,6 +625,8 @@ main(int argc, char **argv)
                 return TFAIL;
         }
 
+	c.value = af_enabled;
+
         if ((fd_v4l = open(v4l_device, O_RDWR, 0)) < 0) {
                 printf("Unable to open %s\n", v4l_device);
                 return TFAIL;
@@ -719,6 +729,11 @@ main(int argc, char **argv)
 		}
 	}
 
+        if (ioctl(fd_v4l, VIDIOC_S_CTRL, &c) < 0)
+        {
+                printf("Warning! VIDIOC_S_CTRL failed Focus Control not available!!\n");
+        }
+
         if (!g_overlay) {
 	        g_alpha.alpha = 255;
 	        g_alpha.enable = 1;
-- 
2.7.4

